### DISCUSSION TITLE
Is it impossible to get notifed as soon as a new window is created?

### DISCUSSION THREAD

**Discussion by WantToLearnJapanese** [Kalmat/PyWinCtl#61] [Q&A]
#58 I got answered that I need to call `getAllWindows` periodically.
Maybe I have no knowledge about the structure of Windows's event system.
But I think the current implementation is not efficient.

Can't we hook some API so that we can get notified as soon as a new window is created?

**Comment by Kalmat** [comment]
Hi!

In Windows, I think it's possible, but not using Python. I think to remember you had to inject a hook to an OS process, using C++, and I believe it has to be done from a separate DLL. Take a look to [this ](https://stackoverflow.com/questions/1024756/how-can-i-be-notified-when-a-new-window-is-created-on-win32) and [this](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexa?redirectedfrom=MSDN).

In Linux and macOS, it's definitely not possible. At least not that I know.

Using `getAllWindows()` is a feasible, pythonic, multi-platform workaround. It's also true that this approach has two potential drawbacks:

1. It may interfere with the main process
2. Notifications will not be exactly immediate

To avoid point 1, create a separate thread which will be in charge of watching windows in a loop, passing a callback that the thread may invoke to notify main process when a new window is created.

To improve point 2, reduce the wait time of the thread loop (in my system, getAllWindows() takes 0.0014870166778564453 seconds, although it highly depends on your system specs and how many open windows you have).

Take a look to this (it's a quick and dirty adaptation of something similar I have developed to control monitors plugged to / unplugged from system):

```
import threading
import time

import pywinctl as pwc


class CheckWindows:

    def __init__(self, callback):
        self.worker = CheckWindowsWorker(callback)
        self.worker.daemon = True
        self.worker.start()

    def stop(self):
        self.worker.stop()
        self.worker.join()


class CheckWindowsWorker(threading.Thread):

    def __init__(self, callback):
        threading.Thread.__init__(self)

        self.callback = callback
        self.windows = [w.getHandle() for w in pwc.getAllWindows()]
        self.kill = threading.Event()
        self.interval = 0.1

    def run(self):

        while not self.kill.is_set():
            windows = [w.getHandle() for w in pwc.getAllWindows()]
            newWindows = [w for w in windows if w not in self.windows]
            if newWindows:
                print("NEW WINDOWS FOUND")
                self.callback(newWindows)
                self.windows = windows
            self.kill.wait(self.interval)

    def stop(self):
        self.kill.set()


def newWinNotification(newWindows):
    for handle in newWindows:
        print("NEW WINDOW", handle, pwc.Window(handle).title)


checkWindows = CheckWindows(newWinNotification)
while True:
    try:
        time.sleep(1)
    except KeyboardInterrupt:
        break
checkWindows.stop()
```

This script reacts to new windows created in up to 0.1 seconds, and is consuming 0.3% of my CPU.
